### ISSUE TITLE
[BUG] for i, value in enumerate(vector[PyObject*]) fails to build

### ISSUE THREAD

**Issue by maxbachmann** [cython/cython#4072]
**Describe the bug**
`for i, value in enumerate(vector[PyObject*])` fails with the error

```
Cannot convert 'vector[PyObject *]' to Python object
```
this works fine with vectors of different types like e.g. `int`

**To Reproduce**
Code to reproduce the behaviour:
```cython
cdef test():
    cdef vector[PyObject*] vec
    cdef size_t i
    for i, value in enumerate(vec):
        pass
```

**Expected behavior**
`value` should be inferred as as `PyObject*`

**Environment:**
 - Cython version master version #f29ef5b50736ecc03f62faf91f428a889c61e1ef

**Comment by da-woods** [comment]
I'm not completely convinced that this is wrong. Although Cython optimizes `enumerate` it's still a Python function requiring Python arguments.

I could definitely see the value in allowing it to be helpful, but I'm not sure we're obliged to.

**Comment by maxbachmann** [comment]
Yes maybe it falls more in the enhancement category. Do you know in which cases it works already?
E.g.
```python
cdef test():
    cdef vector[int] vec
    cdef size_t i
    for i, value in enumerate(vec):
        pass
```
or
```
ctypedef struct my_struct:
    int a
    int b

cdef test():
    cdef vector[my_struct] vec
    cdef size_t i
    for i, value in enumerate(vec):
        pass
```
work. It appears like it fails the moment one of the elements is a pointer.
